src/number_theory/modular_forms/basic.lean

```diff
@@ -0,0 +1,339 @@
+
+@[priority 100]
+instance modular_form_class.modular_form : modular_form_class (modular_form Γ k) Γ k :=
+{ coe := modular_form.to_fun,
+  coe_injective' := λ f g h, by cases f; cases g; congr',
+  slash_action_eq := modular_form.slash_action_eq',
+  hol := modular_form.hol',
+  bdd_at_infty := modular_form.bdd_at_infty' }
+
+@[priority 100]
+instance cusp_form_class.cusp_form : cusp_form_class (cusp_form Γ k) Γ k :=
+{ coe := cusp_form.to_fun,
+  coe_injective' := λ f g h, by cases f; cases g; congr',
+  slash_action_eq := cusp_form.slash_action_eq',
+  hol := cusp_form.hol',
+  zero_at_infty := cusp_form.zero_at_infty' }
+
+variables {F Γ k}
+
+@[simp] lemma modular_form_to_fun_eq_coe {f : modular_form Γ k} : f.to_fun = (f : ℍ → ℂ) := rfl
+@[simp] lemma cusp_form_to_fun_eq_coe {f : cusp_form Γ k} : f.to_fun = (f : ℍ → ℂ) := rfl
+
+@[ext] theorem modular_form.ext {f g : modular_form Γ k} (h : ∀ x, f x = g x) : f = g :=
+fun_like.ext f g h
+
+@[ext] theorem cusp_form.ext {f g : cusp_form Γ k} (h : ∀ x, f x = g x) : f = g :=
+fun_like.ext f g h
+
+/-- Copy of a `modular_form` with a new `to_fun` equal to the old one. Useful to fix
+definitional equalities. -/
+protected def modular_form.copy (f : modular_form Γ k) (f' : ℍ → ℂ) (h : f' = ⇑f) :
+  modular_form Γ k :=
+{ to_fun := f',
+  slash_action_eq' := h.symm ▸ f.slash_action_eq',
+  hol' := h.symm ▸ f.hol',
+  bdd_at_infty' := λ A, h.symm ▸ f.bdd_at_infty' A }
+
+/-- Copy of a `cusp_form` with a new `to_fun` equal to the old one. Useful to fix
+definitional equalities. -/
+protected def cusp_form.copy (f : cusp_form Γ k) (f' : ℍ → ℂ) (h : f' = ⇑f) :
+  cusp_form Γ k :=
+{ to_fun := f',
+  slash_action_eq' := h.symm ▸ f.slash_action_eq',
+  hol' := h.symm ▸ f.hol',
+  zero_at_infty' := λ A, h.symm ▸ f.zero_at_infty' A }
+
+end modular_form
+
+namespace modular_form
+
+open slash_invariant_form
+
+variables {F : Type*} {Γ : subgroup SL(2, ℤ)} {k : ℤ}
+
+instance has_add : has_add (modular_form Γ k) :=
+{ add := λ f g,
+  { hol' := f.hol'.add g.hol',
+    bdd_at_infty' := λ A, (congr_arg is_bounded_at_im_infty (slash_add k A f g)).mpr
+      ((f.bdd_at_infty' A).add (g.bdd_at_infty' A)),
+    .. (f : slash_invariant_form Γ k) + g } }
+
+@[simp] lemma coe_add (f g : modular_form Γ k) : ⇑(f + g) = f + g := rfl
+
+@[simp] lemma add_apply (f g : modular_form Γ k) (z : ℍ) : (f + g) z = f z + g z := rfl
+
+instance has_zero : has_zero (modular_form Γ k) :=
+{ zero := ⟨0, slash_action.mul_zero _, (λ _, mdifferentiable_at_const 𝓘(ℂ, ℂ) 𝓘(ℂ, ℂ)),
+  by {intro a,
+    convert (bounded_at_im_infty_subalgebra ℂ).zero_mem',
+    apply slash_action.mul_zero _ }⟩}
+
+instance has_csmul : has_smul ℂ (modular_form Γ k) :=
+⟨ λ c f,
+  { to_fun := c • f,
+    hol' :=  f.hol'.const_smul c,
+    bdd_at_infty' := λ A, (congr_arg is_bounded_at_im_infty (slash_action.smul_action k A f c)).mpr
+      ((f.bdd_at_infty' A).const_smul_left c),
+     .. c • (f : slash_invariant_form Γ k) }⟩
```

Comment:
I think most of these algebraic instances should be re-worked a bit.  I'll discuss this one as an example but the same applies for all of them (add, zero, copy, mul, ...; modular, cusp); please change all of them.

Firstly, you don't need a `to_fun` field; the `to_fun` information is provided by the `.. c • (f : slash_invariant_form Γ k)`.

Secondly, I think the `bdd_at_infty'` proof will be clearer if it's written as 
```lean
bdd_at_infty' := λ A, by simpa using (f.bdd_at_infty' A).const_smul_left c,
```
This currently fails for two reasons:
(1) `upper_half_plane.bounded_mem` is marked as `@[simp]` but I don't think it should be -- usually one wouldn't want to unfold it.  Can you please remove the simp attribute and put it in manually in any existing proof that actually uses it?
(2) `smul_slash` is not a simp-lemma.  I think it would be safe to make it one; can you let me know if this causes any problems?  But even if it doesn't work to make it one globally, you can say `by simpa [smul_slash] using ...`

Reply:
Providing `to_fun` again is often a good idea as it means it unfolds more nicely.

Reply:
Ok so I got the proof working here with a `simpa`. I'm not sure what it best to do about the `to_fun`'s shall we just leave them if it means things unfold more nicely?

Reply:
Ok I think this is done. I've left the `to_fun`s for now.

Reply:
I think there's still some work to be done here.  One thing the proofs are exposing is that there are some missing projection-notation-form lemmas for `bounded_at_filter` and `zero_at_filter`.  (You have some of the content in proving submodule/submonoid instances, but these are less user-friendly.). I took the liberty of making a quick pass adding these and adjusting the proofs to use them, see [this commit](https://github.com/leanprover-community/mathlib/commit/639b6bd774b14d07f4d7357c1531fd178d1c870e).  If you are happy with that you can add it to your branch using `git cherry-pick 639b6bd774b14d07f4d7357c1531fd178d1c870e`.

There is still some work to be done on the various `simpa [some_lemma] using ...`: can you in each case test whether the `some_lemma` is safe to add as a simp-lemma, and if so, make it one and shorten the proof?

Reply:
> You have some of the content in proving submodule/submonoid instances, but these are less user-friendly.

note: I'm actually not sure these submodule/submonoid instances are useful at all, but let's wait til there's a whole working theory of modular forms, including the graded ring instance, before deleting them!